Make PyContainedProxyBase inherit directly from AbstractPyProxyBase.

jamadden opened the pull request:
This should fix the same problems mentioned in zopefoundation/zope.proxy#5, plus reduce redundant code.

It needs zope.proxy >= 4.1.5, but there's no direct dependency on zope.proxy here, it's inherited from zope.security (I think) so I wasn't sure whether to add a pin here or not.

I thought this needed zopefoundation/persistent#20 to be released, but I just locally ran tox and it passed, so maybe it doesn't need it for the tests. But I think it will for actual use.

mgedmin commented:
> It needs zope.proxy >= 4.1.5, but there's no direct dependency on zope.proxy here, it's inherited from zope.security (I think) so I wasn't sure whether to add a pin here or not.

My rule of thumb: if it imports code from zope.proxy ([and it does](https://github.com/zopefoundation/zope.container/blob/8e608489af600f1acfc63cc90f7c4613bd4bc002/src/zope/container/_proxy.py#L21)), then setup.py needs to list zope.proxy in `install_requires`.  Relying on transitional dependencies is Mildly Wrong.

jamadden commented:
Thanks @mgedmin, that's how we manage our projects too, I just wasn't sure what the convention was here. I added the missing requirements for `zope.proxy` and `persistent`, both of which need specific versions to function under PyPy.

It turns out that the tests that exercise persistent integration (the reason for the pin) weren't being run by tox or travis, so I also added a couple tox environments to run them, and ran one of them on travis too. This revealed a bug in this PR (ContainedProxy needs to be weekly-referenceable) that I fixed.

tseaver commented:
Thanks!

tseaver commented:
Released w/ https://pypi.python.org/pypi/zope.container/4.1.0
